### PR TITLE
Memleak fix

### PR DESCRIPTION
**Pull request by bzed**
My idea of a fix for #222 - works well so far.

### CODE REVIEW

**Comment by AnderEnder** [comment]
@bzed 
Tested on real production traffic. Confirmed that memory leaks are gone and stable RES memory is around 407MB.

**Comment by Dieterbe** [comment]
Hey Bernd and others, just fyi i don't have time right now to look into this now. but hopefully next week or after.

**Comment by bzed** [comment]
Don't worry, happily running with 600mb ram now. Happy holidays! 

Am 22. Dezember 2017 17:13:59 MEZ schrieb Dieter Plaetinck <notifications@github.com>:
>Hey Bernd and others, just fyi i don't have time right now to look into
>this now. but hopefully next week or after.
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/graphite-ng/carbon-relay-ng/pull/248#issuecomment-353628633

-- 
Diese Nachricht wurde von meinem Android-Gerät mit K-9 Mail gesendet.

**Comment by AnderEnder** [comment]
@Dieterbe, any news on this PR?

File: destination/keepsafe.go [hunk 1]
```diff
@@ -43,8 +43,8 @@ func (k *keepSafe) Add(buf []byte) {
 func (k *keepSafe) GetAll() [][]byte {
 	k.Lock()
 	ret := append(k.safeOld, k.safeRecent...)
-	k.safeOld = make([][]byte, 0, k.initialCap)
-	k.safeRecent = make([][]byte, 0, k.initialCap)
+	k.safeOld = k.safeOld[:0]
```
Comment (Dieterbe) [inline destination/keepsafe.go]: this is unsafe because it will store data into the same backing array as the one which is being returned which will be read from by the caller of GetAll().
what's the goal of these lines changing? just a memory-reuse optimization I guess, but doesn't look related to the mem leak itself.

File: destination/keepsafe.go [hunk 2]
```diff
@@ -29,7 +29,7 @@ func (k *keepSafe) keepClean() {
 	for range tick.C {
 		k.Lock()
 		k.safeOld = k.safeRecent
-		k.safeRecent = make([][]byte, 0, k.initialCap)
+		k.safeRecent = k.safeRecent[:0]
```
Comment (Dieterbe) [inline destination/keepsafe.go]: unsafe because now safeRecent and safeOld are backed by the same array

**Comment by Dieterbe** [comment]
Sorry, this code is not good. see #222 for updates and better fix.
